### DISCUSSION TITLE
Is it possible to do a reachability study in a volume ?

### DISCUSSION THREAD

**Discussion by vicltz** [ros-industrial/reach#53] [Q&A]
Thank you for providing this package to the community, it is very powerful and useful.

I have a question regarding the possibilities offered by the package.

Do you think it might be possible to study reachability in a volume ? Instead of having a mesh and generating a point cloud on it, would it be possible to have just a point cloud in a volume and evaluate reachability in that point cloud ?
The idea I had to do it with a mesh is to have a vertical plane mesh, thin on the x axis, and doing the reachability study with this plane multiple times with a position further and further from the robot, and then compile all the results to make a volume.

Does it make sense according to you ?

One other question I had is a more tricky one. I would like to study the reachability of a mobile manipulator that has two arms. Do you think it's possible with the package, and how ? Otherwise, do you think it's possible to do some tricks with the code in order to make this study possible ?

Thank you for your help.

**Comment by marip8** [comment]
> Do you think it might be possible to study reachability in a volume ? Instead of having a mesh and generating a point cloud on it, would it be possible to have just a point cloud in a volume and evaluate reachability in that point cloud ?

Definitely. The function of the [`TargetPoseGenerator` interface](https://github.com/ros-industrial/reach/blob/71eccfc5e44514660c2ca244f374736e89ee8d48/include/reach/interfaces/target_pose_generator.h) is to allow the user to customize the generation of target waypoints for the reach study. The current implementations read point/normals from a .pcd file, but you could implement your own plugin to do this in the way you describe. See [here](https://github.com/ros-industrial/reach/blob/71eccfc5e44514660c2ca244f374736e89ee8d48/src/plugins/point_cloud_target_pose_generator.cpp) and [here](https://github.com/ros-industrial/reach/blob/71eccfc5e44514660c2ca244f374736e89ee8d48/src/plugins/plugins.cpp#L14) for an example in c++ and [here for an example in Python](https://github.com/ros-industrial/reach/blob/71eccfc5e44514660c2ca244f374736e89ee8d48/scripts/test_reach.py#L22).

> One other question I had is a more tricky one. I would like to study the reachability of a mobile manipulator that has two arms. Do you think it's possible with the package, and how ? Otherwise, do you think it's possible to do some tricks with the code in order to make this study possible ?

Yes, with some caveats; I've done this before with [this robot](https://github.com/ros-industrial/reach/blob/71eccfc5e44514660c2ca244f374736e89ee8d48/docs/reach_study.png). To use `reach` as-is, you'll need to:
- define a URDF with two prismatic joints (along x, y) and a revolute joint (about z) to move the mobile base
- define a motion group that includes these 3 joints and the joints of one of your robot arm(s) and an IK solver that can solve
  - You might be able to define a motion group that includes both arms of the robot, but you may not find an IK solver that knows how to resolve both arms for a single target position (thinking specifically about MoveIt). You'll probably have to define two motion groups, where each includes only one of the arms, run two reach studies, and manually combine/compare the results.

FWIW, I'm running a workshop at ROSCon 23 where I plan to talk in more detail about customizing `reach` for use-cases that look significantly different than the one example in the repository. As a part of prep for that workshop, I can add some more documentation about plugin customization and additional examples, so keep an eye out for that and feel free to make suggestions for content

**Reply by saikishor** [reply]
I am also looking for a similar use case to study our Robot. You pointed out on performing 2 reach studies separately and then manually combining/comparing the results. However, I'm also interested in the reachability study of the same point by both manipulators at the same time, there might be instances where this is a very interesting reachability study. 

The problem with the current approach is that when you combine both results, there might be some points that cannot be reached by both arms at the same time as they might collide with each other right? And of course, these configurations will be few. I would like to know if it is possible to do the study considering both at the same time.

**Reply by saikishor** [reply]
Hello @marip8!

Do you think it is possible to perform the study with multiple arms at the same time using reach?. If so, How can this be achieved?. Could you explain in a few lines, so I can try it out?

I'm interested in a study where I want to access the reachability of certain points with 2 arms or more and of course with different approach orientations. As I've explained earlier, accessing them separately might not consider some self-collision case between the arms.

**Reply by marip8** [reply]
If you don't want to combine the results of two different reach studies, I would create a custom IK solver that basically solves reachability for both arms individually and reports the combined result (e.g., boolean and, boolean or etc). If you are using MoveIt, this should be relatively straightforward modification to the [existing MoveIt IK solver](https://github.com/ros-industrial/reach_ros/blob/246b2b003b56484426bf3ad3c74e2a697e088d40/src/ik/moveit_ik_solver.cpp) (add a second motion group name and set an arbitrary position for the other arm to avoid self-collision).

It may be difficult to update the rest of the reach study to accommodate two individual arms though. For example, you would have to consider how to store the pose of the robot in the internal data structure for use by the evaluators that generate a score for each robot pose. It currently only stores one set of robot joints and gives those to the evaluator(s) for scoring. Also the display only knows how to animate one set of robot joints.

Maybe a better solution for running a reach study with two independent arms is to pretend like it's one single large motion group (i.e, with 2x degrees of freedom) and customize the IK solver, evaluator, and display plugins to extract the correct joint indices for each arm and combine the results appropriately
